modules/currency-l0/src/main/scala/org/tessellation/currency/l0/snapshot/services/StateChannelBinarySender.scala

```diff
@@ -0,0 +1,178 @@
+        currentOrdinal <- lastGlobalSnapshotStorage.getOrdinal
+
+        (confirmed, pending) <- pendingR.modify { pending =>
+          val (confirmed, unconfirmed) = pending.zipWithIndex.collect {
+            case (TrackedBinary(tracked, _), cuttingPoint) if confirmedHashesInGlobalSnapshot.contains(tracked.hash) => cuttingPoint + 1
+          }.lastOption
+            .map(pending.splitAt)
+            .getOrElse((pending, List.empty[TrackedBinary]))
+
+          (unconfirmed, (confirmed, unconfirmed))
+        }
+
+        expectedOrdinalsToConfirm <- updateAvgExpectedOrdinalsToConfirm(confirmed, globalSnapshot.ordinal)
+
+        _ <- catchUpMode.update(prev =>
+          if (pending.isEmpty) {
+            false
+          } else {
+            val hasStalled = pending.exists { p =>
+              val ordinalsWithoutConfirmationCount =
+                NonNegLong
+                  .from(currentOrdinal.getOrElse(SnapshotOrdinal.MinValue).value - p.sentAtOrdinal.value)
+                  .getOrElse(NonNegLong.MaxValue)
+              ordinalsWithoutConfirmationCount >= expectedOrdinalsToConfirm
+            }
+            hasStalled || prev
+          }
+        )
+
+        _ <- avgConfirmedPerSnapshotR.update(prev => NonNegLong.from(Math.round((prev.toDouble + confirmed.length) / 2)).getOrElse(prev))
+
+        _ <- logger
+          .warn(
+            s"${pending.length} state channel binaries are still not confirmed on the global state. Waiting for the next global snapshot."
+          )
+          .whenA(pending.nonEmpty)
+      } yield ()
+
+      def processPending: F[Unit] = {
+        val retry = for {
+          pending <- pendingR.get
+          avgConfirmedPerSnapshot <- avgConfirmedPerSnapshotR.get
+          cap = Math.round(avgConfirmedPerSnapshot + (avgConfirmedPerSnapshot.toDouble / 2)).toInt
+          cappedToRetry = pending.take(cap)
+          _ <- cappedToRetry.traverse(tracked => enqueue(tracked.binary))
+        } yield ()
+
+        catchUpMode.get.ifM(retry, Applicative[F].unit)
+      }
+
+      def clearPending: F[Unit] = pendingR.set(List.empty)
+
+      private def updateAvgExpectedOrdinalsToConfirm(
+        confirmedTracked: List[TrackedBinary],
+        confirmedAtOrdinal: SnapshotOrdinal
+      ): F[PosLong] = {
+        val ordinalsPassed = confirmedTracked.map(tb => confirmedAtOrdinal.value - tb.sentAtOrdinal.value)
+        val avgOrdinalsPassed = PosLong.unsafeFrom(Math.round(ordinalsPassed.sum.toDouble / ordinalsPassed.length))
+        expectedOrdinalsToConfirmR.set(avgOrdinalsPassed).as(avgOrdinalsPassed)
+      }
+
+      private def enqueue(binary: Hashed[StateChannelSnapshotBinary]): F[Unit] =
+        for {
+          currentOrdinal <- lastGlobalSnapshotStorage.getOrdinal.map(_.getOrElse(SnapshotOrdinal.MinValue))
+          _ <- pendingR.update { pending =>
+            if (pending.exists(_.binary.hash === binary.hash)) pending
+            else pending :+ TrackedBinary(binary, currentOrdinal)
```

Comment:
Is `:+` (append) the best operation here? Or maybe change the data structure? Given we're appending on a `List` that means each append is _O(n)_ and it makes a copy of the list.

Reply:
Fair point. We need to append elements and either treat it as LIFO or FIFO (with dequeueN) depending on `catchUpMode` so I think we can switch to `Queue` as it has a constant time for append.